### PR TITLE
Timestamp for validators

### PR DESCRIPTION
**Pull request by eesmerdag**

### CODE REVIEW

File: api/skale/delegation.go [hunk 1]
```diff
@@ -57,7 +57,7 @@ func (c *Caller) GetDelegation(ctx context.Context, bc *bind.BoundContract, bloc
 	}
 
 	createT := results[4].(*big.Int)
-	dID := delegationID
+	dID, _ := new(big.Int).SetString(delegationID.String(), 10)
```
Comment (lukanus) [inline api/skale/delegation.go:60]:
```dg := structs.Delegation{
   DelegationID: new(big.Int).Set(delegationID)
```
no need to create another var

File: api/skale/validator.go [hunk 2]
```diff
@@ -57,7 +57,7 @@ func (c *Caller) GetValidator(ctx context.Context, bc *bind.BoundContract, block
 
 	vr := &ValidatorRaw{}
 	vraw := *abi.ConvertType(results[0], vr).(*ValidatorRaw)
-	vID := validatorID
+	vID, _ := new(big.Int).SetString(validatorID.String(), 10)
 	return structs.Validator{
 		ValidatorID:             vID,
```
Comment (lukanus) [inline api/skale/validator.go:62]: same as delegations

File: api/skale/node.go [hunk 3]
```diff
@@ -158,7 +158,7 @@ func (c *Caller) GetNode(ctx context.Context, bc *bind.BoundContract, blockNumbe
 	lrDate := results[5].(*big.Int)
 	IP := results[1].([4]byte)
 	publicIP := results[1].([4]byte)
-	nID := nodeID
+	nID, _ := new(big.Int).SetString(nodeID.String(), 10)
```
Comment (lukanus) [inline api/skale/node.go:161]: same as delegations

File: scraper/structs/query_params.go [hunk 4]
```diff
@@ -45,6 +45,7 @@ type ValidatorStatisticsParams struct {
 	ValidatorID string
 	Type        StatisticTypeVS
 	BlockHeight uint64
+	Time        time.Time
```
Comment (lukanus) [inline scraper/structs/query_params.go:48]: we have two, already, what is third?

File: store/postgresql/validator_statistics.go [hunk 5]
```diff
@@ -6,30 +6,31 @@ import (
 	"math/big"
 	"strconv"
 	"strings"
+	"time"
 
 	"github.com/figment-networks/skale-indexer/scraper/structs"
 )
 
-func (d *Driver) SaveValidatorStatistic(ctx context.Context, validatorID *big.Int, blockHeight uint64, statisticsType structs.StatisticTypeVS, amount *big.Int) (err error) {
+func (d *Driver) SaveValidatorStatistic(ctx context.Context, validatorID *big.Int, blockHeight uint64, time time.Time, statisticsType structs.StatisticTypeVS, amount *big.Int) (err error) {
```
Comment (lukanus) [inline store/postgresql/validator_statistics.go:14]: Don't overwrite package names

File: store/store.go [hunk 6]
```diff
@@ -121,8 +121,8 @@ func (s *Store) CalculateLinkedNodes(ctx context.Context, params structs.Validat
 	return s.driver.CalculateLinkedNodes(ctx, params)
 }
 
-func (s *Store) SaveValidatorStatistic(ctx context.Context, validatorID *big.Int, blockHeight uint64, statisticsType structs.StatisticTypeVS, amount *big.Int) (err error) {
-	return s.driver.SaveValidatorStatistic(ctx, validatorID, blockHeight, statisticsType, amount)
+func (s *Store) SaveValidatorStatistic(ctx context.Context, validatorID *big.Int, blockHeight uint64, time time.Time, statisticsType structs.StatisticTypeVS, amount *big.Int) (err error) {
```
Comment (lukanus) [inline store/store.go:124]: don't overwrite package names 2

File: store/postgresql/validator_statistics.go [hunk 7]
```diff
@@ -206,14 +207,14 @@ func (d *Driver) CalculateLinkedNodes(ctx context.Context, params structs.Valida
...
+	_, err = tx.ExecContext(ctx, `INSERT INTO validator_statistics (validator_id, block_height, time, statistic_type, amount)
+									(SELECT  $1, $2, $3, $4, count(*) AS amount
 									FROM nodes
-									WHERE validator_id = $1 AND address != $4
+									WHERE validator_id = $1 AND address != $5
 									GROUP BY validator_id LIMIT 1)
 								ON CONFLICT (validator_id, block_height, statistic_type)
 								DO UPDATE SET amount = EXCLUDED.amount`,
-		params.ValidatorID, params.BlockHeight, structs.ValidatorStatisticsTypeLinkedNodes, zero)
+		params.ValidatorID, params.BlockHeight, params.Time, structs.ValidatorStatisticsTypeLinkedNodes, zero)
```
Comment (lukanus) [inline store/postgresql/validator_statistics.go:217]: zero is literally `AND address != 0` no ?

File: api/skale/delegation.go (hunk 1 above)
Comment (eesmerdag) [reply api/skale/delegation.go:60]: 👍

File: api/skale/validator.go (hunk 2 above)
Comment (eesmerdag) [reply api/skale/validator.go:62]: 👍

File: api/skale/node.go (hunk 3 above)
Comment (eesmerdag) [reply api/skale/node.go:161]: 👍

File: store/postgresql/validator_statistics.go (hunk 7 above)
Comment (eesmerdag) [reply store/postgresql/validator_statistics.go:217]: it is same

File: scraper/structs/query_params.go (hunk 4 above)
Comment (eesmerdag) [reply scraper/structs/query_params.go:47]: the third one represents BlockTimeStamp, i will rename it

Comment (lukanus) [reply scraper/structs/query_params.go:47]: but those two other ones are block timestamps as well :)

Comment (eesmerdag) [reply scraper/structs/query_params.go:47]: those are not used at all, i will remove them

Comment (lukanus) [reply scraper/structs/query_params.go:47]: hmmm but sounds like it's needed for hmmm pagination :) and not getting entire database at once :)

File: store/postgresql/validator_statistics.go (hunk 7 above)
Comment (lukanus) [reply store/postgresql/validator_statistics.go:217]: No, passing unnecessary variable that it constant is not the same as just adding 0 to request directly :)

File: scraper/structs/query_params.go (hunk 4 above)
Comment (lukanus) [reply scraper/structs/query_params.go]: We should have a way to range responses in backend. Also it's impossible to efficiently paginate few k records, so we have to enable that, and in requests we should have this

**Review by lukanus — APPROVED** [review]

**Pull request merged**